File: sim/btd/tower.py
"""Tower entity. Mirrors Tower.as."""

from __future__ import annotations

from dataclasses import dataclass

from .constants import TOWER_STATS


@dataclass
class Tower:
    id: int
    type: str
    x: float
    y: float
    attack_rate: int
    attack_radius: float
    shoot_power: float
    pierce_max: int
    spent_on_me: int
    icebreak: bool = False     # bullets can pop frozen bloons
    leadbreak: bool = False    # bullets can pop lead bloons
    is_spread: bool = False    # fires N shards in a fan (tack, ice); no aimed bullet
    is_attacker: bool = True   # False for beacons — skip fire logic, still tick counter
    freeze_len: int = 0        # frames a frozen bloon stays frozen (ice only)
    bullet_scale: float = 1.0  # multiplies bullet.radius and explosion_radius at fire time
    # Upgrade flags. upgrade1 + upgrade3 form path 1; upgrade2 + upgrade4 form
    # path 2. Path-locking lives in BloonsSim.upgrade_tower.
    upgrade1: bool = False
    upgrade2: bool = False
    upgrade3: bool = False
    upgrade4: bool = False
    # Visual / behavioural flags flipped by upgrades.
    transformed: bool = False  # blade tack, glaive boomerang, missile bomb, multishot spikeopult
    laser: bool = False        # super monkey laser (upgrade3) / plasma (upgrade4)
    # Set by the beacon-buff refresh each frame. Read by _acquire_target /
    # _tick_towers when deciding effective range / rate.
    beacon_radius_active: bool = False
    beacon_rate_active: bool = False
    pop_count: int = 0
    time_since_last_shot: int = 0
    # Targeting mode. AS only ever uses "first"; "last" is supported for
    # completeness but no shipped tower switches to it.
    ai_mode: str = "first"

    @classmethod
    def from_type(cls, tower_id: int, type_: str, x: float, y: float) -> "Tower":
        stats = TOWER_STATS[type_]
        return cls(
            id=tower_id,
            type=type_,
            x=x,
            y=y,
            attack_rate=stats["attackRate"],
            attack_radius=stats["attackRadius"],
            shoot_power=stats["shootPower"],
            pierce_max=stats["pierceMax"],
            spent_on_me=stats["cost"],
            icebreak=stats.get("icebreak", False),
            leadbreak=stats.get("leadbreak", False),
            is_spread=stats.get("is_spread", False),
            is_attacker=stats.get("is_attacker", True),
            freeze_len=stats.get("freeze_len", 0),
        )

File: sim-rs/src/entity/tower.rs
/// Tower entity. Mirrors Tower.as + sim/btd/tower.py.

use crate::constants::{TowerType, TOWER_STATS};

#[derive(Clone, Debug)]
pub(crate) struct Tower {
    pub id: u32,
    pub type_: TowerType,
    pub x: f64,
    pub y: f64,
    pub attack_rate: u16,
    pub attack_radius: f64,
    pub shoot_power: f64,
    pub pierce_max: u16,
    pub spent_on_me: u32,
    pub icebreak: bool,
    pub leadbreak: bool,
    pub is_spread: bool,
    pub is_attacker: bool,
    pub freeze_len: u16,
    pub bullet_scale: f64,
    // Upgrade flags.
    pub upgrade1: bool,
    pub upgrade2: bool,
    pub upgrade3: bool,
    pub upgrade4: bool,
    // Visual / behavioural flags.
    pub transformed: bool,
    pub laser: bool,
    // Beacon buffs (refreshed each frame).
    pub beacon_radius_active: bool,
    pub beacon_rate_active: bool,
    // Stats.
    pub pop_count: u32,
    pub time_since_last_shot: u16,
    pub placed_round: u16,
}

impl Tower {
    pub fn from_type(tower_id: u32, type_: TowerType, x: f64, y: f64) -> Self {
        let stats = TOWER_STATS.iter().find(|(t, _)| *t == type_).map(|(_, s)| s).unwrap();
        Self {
            id: tower_id,
            type_,
            x,
            y,
            attack_rate: stats.attack_rate,
            attack_radius: stats.attack_radius,
            shoot_power: stats.shoot_power,
            pierce_max: stats.pierce_max,
            spent_on_me: stats.cost,
            icebreak: stats.icebreak,
            leadbreak: stats.leadbreak,
            is_spread: stats.is_spread,
            is_attacker: stats.is_attacker,
            freeze_len: stats.freeze_len,
            bullet_scale: 1.0,
            upgrade1: false,
            upgrade2: false,
            upgrade3: false,
            upgrade4: false,
            transformed: false,
            laser: false,
            beacon_radius_active: false,
            beacon_rate_active: false,
            pop_count: 0,
            time_since_last_shot: 0,
            placed_round: 0,
        }
    }
}
